lib/api/ldp/ldp-request.js

```diff
@@ -0,0 +1,145 @@
+   * @param options {object}
+   *
+   * @param options.host {SolidHost} Host / config object
+   *
+   * @param options.target {URL} Parsed full request url
+   *
+   * @param options.req {HttpRequest} Save the Express req object, useful
+   *   to access headers/preferences, writeStream for body, etc.
+   *
+   * @param [options.authn=null] {object} Authentication object, contains WebID
+   *   string, as well as any bearer credentials/tokens. Needed for
+   *   authenticated fetch (of remote group ACLs, of Copy resources, etc).
+   *   Null if request is not authenticated.
+   *
+   * @param [options.parsedBody] {?} Request body, parsed when appropriate
+   *   (when content-type is parse-able)
+   */
+  constructor (options) {
+    this.host = options.host
+    this.target = options.target
+    this.req = options.req
+    this.authn = options.authn
+    this.parsedBody = options.parsedBody
+  }
+
+  static from ({req, host}) {
+    let request
+
+    switch (req.method) {
+      case 'head':
+        /**
+         * Checks to see if target exists
+         */
+        request = LdpHeadRequest.from({req, host})
+        break
+      case 'get':
+        /**
+         * Use cases to handle:
+         *  - LdpListContainerRequest, if target is an ldp container
+         *  - However, if Accept: header is HTML, check if index.html exists
+         *  - Need to also support similar use case if index.ttl exists
+         *  - Otherwise, plain LdpGetRequest
+         */
+        request = LdpGetRequest.from({req, host})
+        break
+      case 'put':
+        /**
+         * Creates a resource or container, and any necessary containers above it in
+         * the hierarchy. Idempotent.
+         *
+         * If target is a container:
+         *   - creates the container if none existed
+         *   - does nothing if container exists, does not delete/clear existing contents
+         *   - has `mkdir -p` semantics (creates missing container hierarchy)
+         *
+         * If target is a resource:
+         *   - writes the resource, always overwriting existing contents if any existed
+         *     (since we don't support conditional `If-None-Match` requests)
+         *   - has `mkdir -p` semantics (creates missing container hierarchy)
+         */
+        request = LdpPutRequest.from({req, host})
+        break
+      case 'post':
+        /**
+         * Creates a new resource or container in the target container. The name of
+         * this new resource is derived as follows:
+         *
+         *  - Use contents of `Slug` header, if provided and no resource with same
+         *    name already exists
+         *  - Use contents of Slug plus UUID, if resource with same name exists
+         *  - Generate a UUID if no Slug given
+         *
+         * Does NOT use `mkdir -p` semantics (does not create missing containers)
+         *
+         * Throws:
+         *  - 400 error if target doesn't end in a / (is not a container)
```

Comment:
Actually, this doesn't happen currently (POST doesn't throw a 400 if target path does not end in a /). Since the dispatching between "create container" and "create plain resource" is done via the Link: type rel, the handler is lax about a missing / at the end (in fact, it adds it if it's not there).